### PR TITLE
feat: Add support for message keys in typed producer

### PR DESCRIPTION
**Pull request by rohits1101**
## Pull Request Submission Checklist

Please confirm that you have done the following before requesting reviews:

- [x] I have confirmed that the PR type is appropriate for the change I am making according to
  the [Honest Pull Request and Commit Message Naming Conventions](https://www.notion.so/honestbank/Pull-Request-and-Commit-Message-Naming-Conventions-bd97f2cbb34c4c73b1ff3a3e384b850c).
- [x] I have typed an adequate description that explains **why** I am making this change.
- [x] I have installed and run standard pre-commit hooks that lints and validates my code.

### Description

* The kafka [producer API](https://github.com/honestbank/kp/blob/main/v2/producer/producer.go#L20) does not support setting message keys
* In our kafka-infrastructure, by default the topics are set to compat on keys
* Having empty keys deletes the messages on the topic
* To solve this, we need to do one of the two:
  1. Change the topic compaction policy (to not delete, to not compact on message keys)
  2. Change the KP interface to produce the messages with keys
* We can go ahead with (2) because it also aligns with the organization wide convention of (1) topic deletion policy, (2) messages between acq, and mlops having their keys as application-status-ids

### AC

* Update the [producer interface](https://github.com/honestbank/kp/blob/main/v2/producer/producer.go#L20) to also accept the keys, and produce the messages with message keys

- - -
This change is [Reviewable](https://reviewable.io/reviews/honestbank/kp/88)

### CODE REVIEW

**Review by kelu-diao — APPROVED** [review]

File: v2/producer/producer.go [hunk 1]
```diff
@@ -18,22 +18,36 @@ type producer[BodyType any] struct {
...
 		Partition: kafka.PartitionAny,
 	}
+
 	msg := &kafka.Message{
 		TopicPartition: partition,
 		Value:          value,
 	}
+
+	if key != nil {
+		msg.Key = key
```
Comment (misranitin) [inline v2/producer/producer.go:45]: If the default value of `msg.Key` is `nil`, then can we skip checking if key is `nil`?

File: docker-compose.integration.yaml [hunk 2]
```diff
@@ -17,7 +17,7 @@ services:
     networks:
       kafka:
   local-kafka1:
-    image: confluentinc/cp-kafka:5.5.5
+    image: confluentinc/cp-kafka:7.3.2
```
Comment (onkarbanerjee) [inline docker-compose.integration.yaml:20]: This file is supposed to be synced from .github as mentioned in the comment above. Can we follow the same?

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -43,19 +43,19 @@ jobs:
         run: docker compose up -d
       - name: wait for kafka
         run: |
-          wget https://downloads.apache.org/kafka/3.2.3/kafka_2.13-3.2.3.tgz
-          tar -xzf kafka_2.13-3.2.3.tgz
-          ./kafka_2.13-3.2.3/bin/kafka-topics.sh --bootstrap-server localhost:9092 --list
-          ./kafka_2.13-3.2.3/bin/kafka-topics.sh --bootstrap-server localhost:9092 --list
+          wget https://downloads.apache.org/kafka/3.4.1/kafka_2.13-3.4.1.tgz
```
Comment (onkarbanerjee) [inline .github/workflows/build.yaml:46]: Not sure, why this file is not synced from  `.github` though

Comment (rohits1101) [reply .github/workflows/build.yaml:46]: Looks like we are performing these extra steps specific to KP. To sync them from workflows, we'd have to copy over these to that repo, and sync from there. Doing that for a single repository doesn't seem worth the effort.

Comment (onkarbanerjee) [reply .github/workflows/build.yaml:46]: In general i feel this file can very much be synced from .github and maybe use one of the existing ones. If we feel like we need separate steps for integration tests then we can use a build file without the int tests from .github and a custom integration_tests.yaml file.

**Review by cyberhck — CHANGES_REQUESTED** [review]
keys aren't supported in typed producer by design, they're supposed to be done through middlewares (they're work in progress in #82 ) please make that ready instead.

**Review by onkarbanerjee — APPROVED** [review]

**Pull request closed without merging**